### PR TITLE
fix: Capture as many channel mentions as possible

### PR DESCRIPTION
**Pull request by zevaryx**
## What type of pull request is this?

- [x] Non-breaking code change
- [ ] Breaking code change
- [ ] Documentation change/addition
- [ ] Tests change

## Description

This adds in some extra regex matching during `Message._process_dict` to scan the message content for more channel mentions, and then process them as if they were normal channel mentions. 

This bug is seen here: #324 

While this isn't technically a bug, but rather an API oddity, we still believe that it's necessary to "fix" this for the end users' sake.

## Changes

- Add regex to `Message._process_dict` to capture more channel mentions

## Checklist

- [x] I've formatted my code with [Black](https://black.readthedocs.io/en/stable/)
- [x] I've ensured my code works on `Python 3.10.x`
- [x] I've tested my code

### CODE REVIEW

**Review by LordOfPolls — APPROVED** [review]

File: dis_snek/models/discord/message.py [hunk 1]
```diff
@@ -279,10 +282,20 @@ def _process_dict(cls, data: dict, client: "Snake") -> dict:
...
                 mention_channels.append(ChannelMention.from_dict(channel_data, client))
+            for channel_id in extra:
+                if channel := client.get_channel(channel_id):
+                    channel_data = {
+                        "id": channel.id,
+                        "guild_id": channel._guild_id,
+                        "type": channel.type,
+                        "name": channel.name,
+                    }
+                    mention_channels.append(ChannelMention.from_dict(channel_data, client))
```
Comment (benwoo1110) [inline dis_snek/models/discord/message.py]: hmm will it cause possible duplicated `mention_channels` entries? If it is already in `data["mention_channels"]` and you parsed it again with regex.

File: dis_snek/models/discord/message.py [hunk 2]
```diff
@@ -279,10 +282,20 @@ def _process_dict(cls, data: dict, client: "Snake") -> dict:
                 mention_ids.append(client.cache.place_user_data(user_data).id)
         data["mention_ids"] = mention_ids
 
-        if "mention_channels" in data:
+        extra = channel_mention.findall(data["content"])
+        if "mention_channels" in data or len(extra) > 0:
```
Comment (benwoo1110) [inline dis_snek/models/discord/message.py]: The if logic needs to be separated, or not you will loop through data["mention_channels"] when it is None.

```
Traceback (most recent call last):
  File "c:\users\wben1\documents\github\dis_snek\dis_snek\api\events\processors\message_events.py", line 29, in _on_raw_message_create
    msg = self.cache.place_message_data(event.data)
  File "c:\users\wben1\documents\github\dis_snek\dis_snek\client\smart_cache.py", line 402, in place_message_data
    message = Message.from_dict(data, self._client)
  File "c:\users\wben1\documents\github\dis_snek\dis_snek\models\discord\base.py", line 24, in from_dict
    data = cls._process_dict(data, client)
  File "c:\users\wben1\documents\github\dis_snek\dis_snek\models\discord\message.py", line 288, in _process_dict
    for channel_data in data["mention_channels"]:
KeyError: 'mention_channels'
```

**Review by benwoo1110 — CHANGES_REQUESTED** [review]
Some issues that I spotted after doing some testing: (yes I tested, amazing right?)

File: dis_snek/models/discord/message.py (hunk 2 above)
Comment (zevaryx) [reply dis_snek/models/discord/message.py]: Resolved both in latest commit

**Comment by benwoo1110** [comment]
merge conflict 💀

**Comment by zevaryx** [comment]
> merge conflict 💀

Fixed

File: dis_snek/models/discord/message.py [hunk 3]
```diff
@@ -279,12 +282,25 @@ def _process_dict(cls, data: dict, client: "Snake") -> dict:
                 mention_ids.append(client.cache.place_user_data(user_data).id)
         data["mention_ids"] = mention_ids
 
+        found_ids = []
         if "mention_channels" in data:
             mention_channels = []
```
Comment (benwoo1110) [inline dis_snek/models/discord/message.py]:
```suggestion
        mention_channels = []
        if "mention_channels" in data:
```

```
Traceback (most recent call last):
  File "c:\users\wben1\documents\github\dis_snek\dis_snek\api\events\processors\message_events.py", line 29, in _on_raw_message_create
    msg = self.cache.place_message_data(event.data)
  File "c:\users\wben1\documents\github\dis_snek\dis_snek\client\smart_cache.py", line 404, in place_message_data
    message = Message.from_dict(data, self._client)
  File "c:\users\wben1\documents\github\dis_snek\dis_snek\models\discord\base.py", line 24, in from_dict
    data = cls._process_dict(data, client)
  File "c:\users\wben1\documents\github\dis_snek\dis_snek\models\discord\message.py", line 302, in _process_dict
    mention_channels.append(ChannelMention.from_dict(channel_data, client))
UnboundLocalError: local variable 'mention_channels' referenced before assignment
```

**Review by benwoo1110 — CHANGES_REQUESTED** [review]

File: dis_snek/models/discord/message.py (hunk 3 above)
Comment (zevaryx) [reply dis_snek/models/discord/message.py]: Whoops, fixed

**Review by benwoo1110 — APPROVED** [review]
ok finally looks good (I got lazy to suggest change and just directly push a issue I found)

**Pull request merged**